### ISSUE TITLE
[mig modules] Support bytes string search in File module (Bugzilla #1163598)

### ISSUE THREAD

**Issue by jvehent** [mozilla/mig#120]
Migrated from https://bugzilla.mozilla.org/show_bug.cgi?id=1163598
Assigned to: Julien Vehent [:ulfr]

On 2015-05-11 06:57:45 -0700, Julien Vehent [:ulfr] wrote:

> The file module should allow searching for files that contain a specific hexadecimal encoded bytes string.
> 
> This feature is implemented in the memory module under the `-bytes` options, we should reuse the name in the file module.
> 
> The implementation should use a sliding window to match bytes string at the edge of search buffers, and not using line by line search as is currently the case for regexp searches.

**Comment by tydavis** [comment]
I think I have the basics so far for a `%sbytes` parameter and building in a check for valid hex/bytes search. That appears to be the easy part -- now I have to implement a checkBytes function equivalent to checkContents -  https://github.com/mozilla/mig/blob/master/modules/file/file.go#L1227

Not sure if I'm on the right track and would appreciate any input.

**Comment by ameihm0912** [comment]
This seems like the right approach to me. I think we'd want a new function as you suggest, since `checkContents` does a few extra things like applying macroal (match all content regex on all lines) which we wouldn't want with a byte string search.

`checkBytes` should be simpler but we might want to look at implementing some sort of overlapping search (e.g., read a block of bytes from the file, check it for the byte string, then move up a certain distance so the next search includes part of the previous) since we will not want to read the entire file into memory.

**Comment by tydavis** [comment]
@ameihm0912 Agreed -- I would use that sliding-window design similar to https://gist.github.com/tydavis/8d4e80c37ae096b4a062baba5c54532f -- though, for Bytes, instead of strings.

@jvehent Thanks for the link -- it looks like that uses multiple buffers. Smart idea.

I'm thinking of scanning by reading a blocksize (4k? bigger?), store it in a slice of 2*blocksize (8k?), and set the "sliding window" of search based on the length of the hex key.  When we pass the halfway point by at least two scan lengths through the big block, we delete the , append the new read at the end, and keep going. That way we're always scanning for a possible match, reading in appropriately sized chunks, and never overwriting our current scan.

I hope this makes sense -- I'll whip something up in code and it'll probably be clearer.

**Comment by tydavis** [comment]
Here's where I am so far: https://github.com/tydavis/mig/commit/44cfc97c6740557862cb47be1c8235103cb75091

The design is to read 4k into memory at a time, then run `bytes.Contains(buffer, hexvalue)` along the whole buffer. 

If the file is greater than 4k, we read an additional 4k off the file into a separate buffer, and create a smaller buffer with the number of bytes equal to `2 * len(hexvalue)` from the end of the original buffer. Then we merge the small buffer with the read buffer and continue the loop. This ensures the new buffer always contains enough of the prior data to match a segment that spans two reads.  

This method is slightly inefficient, as it always scans the zone of twice the length of the hex query for each read operation, but I think it's an acceptable tradeoff to accomplish the sliding window design without a significantly more complex algorithm.

Right now, I'm stuck on two things:

1. I don't know exactly how to extract the `search.Bytes` query value out of the search parameters. As far as I can tell, this search will run multiple "searches" per individual file (`fileEntry`) so we're going to end up reading the whole file for each search iteration.  Is that the correct pattern?
2. How do I go about indicating I have found a match?  Set [this variable](https://github.com/mozilla/mig/blob/master/modules/file/file.go#L1405) per check? (`search.check.storeMatch(f.filename)`)

**Comment by ameihm0912** [comment]
@jvehent may be able to provide some insight here, but I believe what you might be missing regarding the first note is you want to loop over the "checks" constructed in `makeChecks`, like:

https://github.com/mozilla/mig/blob/master/modules/file/file.go#L1281

from there I believe you can access your `bytes` value you added in `makeChecks`.

Regarding the second point this seems correct yes.

I'd have to go through this code a bit more in detail, the file module is quite complicated due to it's ability to perform multiple searches at the same time.

**Comment by jvehent** [comment]
/me grabs a headlamp and goes dive into code he hasn't looked at in a couple years

**Comment by jvehent** [comment]
When `pathWalk` traverse the file system, it activates and deactivates individual searches, but doesn't know which `checks` need to run. Then

**Comment by jvehent** [comment]
First, a bit of terminology: a `search` contains a path and a number of check. a `check` apply a specific filter, like `content`, `size`, or `sha256`. So, search is the top-level container, and check is the actual filter.

When `pathWalk` traverses the tree, it activates/deactivates individual searches, but doesn't know which `checks` need to run. Then, `evaluateFile` is executed on each file that has checks:

1. First, it runs the checks that look at file metadata, because those are fast, and will allow us to exit early if matchall is set and the check doesn't match.
2. Then, we run the expensive checks, like `checkContent` and `checkHash`, that will read the entire file and thus are more expensive. You'll want `checkBytes` to be in there too.

When we enter `checkContent`, the function doesn't know which check needs to run. It's entirely possible that no content check applies to this path, so the first thing we do is iterate through all the searches to see which ones are active, then in each search, through all the content checks to see which ones are active. We use a `checkstatus` map to keep a list of checks that need to run on a given file.

Then we read the file, and for each iteration of the scanner loop, we run the active checks. As Aaron said, there's a bit of magic in `checkContent` around the `macroal` mode that you can safely ignore when implementing `checkBytes`. However, you may want to implement inverse matches, that's up to you.

When a match is found, we run `c.storeMatch(f.filename)` and the file is added to the results.

Then, when all the searches are done, `buildResults` takes all the data and creates the JSON that is returned to the agent. You want to add some code there as well.

**Comment by tydavis** [comment]
Thanks @jvehent @ameihm0912 -- I think I've got it. Now I need to build a reliable test for it.

**Comment by tydavis** [comment]
@jvehent @ameihm0912 https://github.com/mozilla/mig/pull/370
